src/llvm-multiversioning.cpp

```diff
@@ -469,6 +471,16 @@ uint32_t CloneCtx::collect_func_info(Function &F)
                     if (name.startswith("llvm.muladd.") || name.startswith("llvm.fma.")) {
                         flag |= JL_TARGET_CLONE_MATH;
                     }
+                    else if (name.startswith("julia.cpu.")) {
+                        if (name.startswith("julia.cpu.have_fma.")) {
+                            // for some platforms we know they always do (or don't) support
+                            // FMA. in those cases we don't need to clone the function.
+                            if (!always_have_fma(*callee).hasValue())
+                                flag |= JL_TARGET_CLONE_CPU;
+                        } else {
+                            flag |= JL_TARGET_CLONE_CPU;
```

Comment:
FMA is in JL_TARGET_CLONE_MATH, so this also may need that flag too, since they work in tandem together

Reply:
But this only matches the `have_fma` intrinsic; presumably the containing function then also contains a call to `fma` which would have set `JL_TARGET_CLONE_MATH`.

Reply:
My question is why you would ever want to clone this call but not clone the `fma` call itself, since that seems like it would cause divergence in the results returned from the have_fma call vs the use of the `fma` algorithm, resulting in incorrect program behaviors.

Reply:
> why you would ever want to clone this call but not clone the fma call itself

You wouldn't, but that also can't happen. The check for `llvm.fma` is right before that, and would set the necessary clone flag: https://github.com/JuliaLang/julia/blob/5864e4341b0f11879a566bb6dbc571e230c26691/src/llvm-multiversioning.cpp#L467-L469

Reply:
The fma call and the have_fma call are probably in different function however